R/md_document.R

```diff
@@ -80,11 +80,22 @@ md_document <- function(variant = "markdown_strict",
       to = variant,
       from = from_rmarkdown(extensions = md_extensions),
       args = args,
-      ext = ext,
-      lua_filters = if (number_sections) pkg_file_lua("number-sections.lua")
+      lua_filters = if (number_sections) pkg_file_lua("number-sections.lua"),
+      ext = ext
     ),
     clean_supporting = FALSE,
     df_print = df_print,
+    pre_processor = if (number_sections) {function(metadata, input_file, ...) {
+      pandoc_convert(
+        input_file, to = "markdown", output = input_file,
+        options = c(
+          "--standalone",
+          "--lua-filter", pkg_file_lua("number-sections.lua"),
+          "--metadata", "preprocess_number_sections=true"
+        )
+      )
+      return(character(0L))
+    }},
```

Comment:
Yes this is because of this.
We set the expected env var in `rmarkdown::render()` - so when runnning the filter in a preprocessor step with pandoc convert, it is not set. 

Maybe it would have been better to set it in `pkg_file_lua()` directly. See long discussion and different choices in https://github.com/rstudio/rmarkdown/pull/2102

Right now, this is missing this to work: 
https://github.com/rstudio/rmarkdown/blob/9036f864d989d06b966c99494521e521b479c700/R/render.R#L884-L888

As documented in the Lua Filters now, this env var needs to be set in order to used the `.lua` files that uses needs to use `shared.lua`. 

We can revisit for sure.

Reply:
Thank you for your insights!!

Well, as @yihui's https://github.com/rstudio/rmarkdown/pull/2102#discussion_r610737639, we need to unset the env var after the filter is run. If we `set_envvar()` in `pkg_file_lua()`, we cannot unset in `pkg_file_lua()`, but in the higher level. This makes codes complicated...

I think I can work around the problem by using `render` instead of `pandoc_convert`. I'll try it.

Reply:
I did not have a close look at your code here yet so I guess you are force to do this two processing. 
One specific case is that you can also set the env in your preprocessor for this specific needs as you are using a filter that needs it. 

Maybe some wrappers a-la **withr** can come handy here: 
```r
with_shared_lua(pandoc_convert(...))
```
would run the command after setting the env and would remove it after. 

We can just also set the env var earlier in `render()` so that it is available when executing the pre processor function. That would be the easiest I think. It would not be set only if there is a lua filter to use though (this is why it is at its place right now) and the env var would be set anytime as soon as `run_pandoc = TRUE`. I think it could fix the PR, isn't it ?

Reply:
Yeah, that is true, and using `render` sounds overkill. Ill do that way later